cmd/server-main.go

```diff
@@ -223,8 +226,57 @@ func serverMain(ctx *cli.Context) {
 	handleSignals()
 }
 
+// Lists all your local mount points belonging to a given prefix.
+func listMountsPrefix(prefix string) ([]mountinfo.MountPoint, error) {
+	mounts, err := mountinfo.MountInfo()
+	if err != nil {
+		// Ignore if mountinfo thinks /proc/mounts is missing
+		// or minio has no permission to read it. In such cases
+		// we just treat it as success and hopefully report
+		// cross-device issues if any at a later stage.
+		if os.IsNotExist(err) || os.IsPermission(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	var newMounts []mountinfo.MountPoint
+	for _, mount := range mounts {
+		if strings.HasPrefix(mount.Path, prefix) {
+			newMounts = append(newMounts, mount)
+		}
+	}
+	return newMounts, nil
+}
+
+// check if path has cross-device mounts.
+func checkPathCrossDeviceMounts(path string) error {
+	mounts, err := listMountsPrefix(path)
+	if err != nil {
+		return err
+	}
+	msg := `Cross-device mounts detected on export path (%s) at following locations %s. Export path should not have any sub-mounts, refusing to start.`
+	for _, mount := range mounts {
+		// At this point after listing all the common mounts there are two cases.
+		// - mount.Path matches exact with `path` means we can proceed no error here.
+		// - mount.Path doesn't match (means cross-device mount), should error out.
+		if mount.Path != path {
+			// if paths didn't match then we do have cross-device mount.
+			err = fmt.Errorf(msg, path, mounts)
+		}
+	}
+	return err
+}
+
 // Initialize object layer with the supplied disks, objectLayer is nil upon any error.
 func newObjectLayer(endpoints EndpointList) (newObject ObjectLayer, err error) {
+	for _, endpoint := range endpoints {
```

Comment:
You could add this check to single endpoint creation ie `NewEndpoint()` should do this validation.